### PR TITLE
feat(networking): add backoff period after failed dial

### PR DESCRIPTION
**Pull request by alrevuelta**
Closes https://github.com/waku-org/nwaku/issues/1414

**Summary:**
After failing to dial a peer, it adds a backoff to it so that we don't attempt to dial the same peer again during some time. The time to wait depends on the amount of consecutive failures, calculated with the following formula:

```
initialBackoffInSec*(backoffFactor^(failedAttempts-1))
```

Note that `initialBackoffInSec` and `backoffFactor` are configurable values that allow to configure how aggressive the backoffs are. Using `initialBackoffInSec=120` and `backoffFactor=4` the times to wait would be:

```
120s, 480s, 1920, 7680s
```

This PR helps nodes to rapidly increase their number of connections, since less time is wasted in trying to connect to nodes that fail. This improvement is even more noticeable in networks whith a high ratio of non reachable peers. Note that this backoff only applies to `relay` peers.

**Changes**:
- [x] Add exponential backoff after a failed dial.
- [x] Some minor test refactoring.
- [x] Some minor logging improvements.
- [x] Directly use `switch.dial` in ping (keep alive)

### CODE REVIEW

**Review by jm-clius — COMMENTED** [review]
Although I agree with the idea of exponential backoffs, I'm not sure I agree that this should be part of (every) `dialPeer`. In my mind, if an "application" (or just a protocol) calls `dialPeer`, it should attempt to connect to that peer immediately without considering delays/backoffs (except maybe a _very_ recent failure). The protocol/application may "know better" than the peer manager about when a peer  has become available again - especially when we're in the 14h backoff period. If a protocol attempts to continuously dial an unreachable peer, there is a problem with that protocol that should be addressed. The only protocol that _currently_ requires continuous connection attempts, however, is Relay. In my mind then there should be a connectivity loop that continuously attempts to connect to some peers from the peer store and respects the backoff period for these peers _within this connectivity loop_. In this approach, `dialPeer` will be available to protocols that want to make an ad-hoc connection to a specific peer and should remain a simple way to attempt to dial the peer (there may be scope for some sanity checks here, but I don't think a protocol should attempt to dial a peer and then wait ~14 hours for the dial to be attempted).

**Comment by jm-clius** [comment]
Of course, similar connectivity guarantees could be useful for service protocols in future too. I just think this connectivity attempts and backoffs should be done in parallel and not serially/ad-hoc whenever attempting to dial.

**Comment by alrevuelta** [comment]
Thanks @jm-clius, fair comments. How about only respecting the backoff for the relay protocol, and having a direct path for the service protocols? Also related to "service protocol slots" https://github.com/waku-org/nwaku/issues/1461.

Some comments:

> The protocol/application may "know better" than the peer manager about when a peer has become available again

mm not sure I follow. Can you provide an example? Not sure how the application layer can have knowledge of this.

> The only protocol that currently requires continuous connection attempts, however, is Relay. In my mind then there should be a connectivity loop that continuously attempts to connect to some peers from the peer store and respects the backoff period for these peers within this connectivity loop

Nice, as stated I will implement that, and have a direct path for service protocols.

> it should attempt to connect to that peer immediately without considering delays/backoffs (except maybe a very recent failure)

Sure. Linking this to the "service protocol slots" https://github.com/waku-org/nwaku/issues/1461. I can implement n retries for "slotted" peers. If its fine will leave that for the PR fixing that issue.

**Comment by alrevuelta** [comment]
Will leave this PR on hold since with the existing code, I can't differenciate between service (store, lp,...) and relay peers. Once I implement this feature https://github.com/waku-org/nwaku/issues/1461 I will be able to know which peers are "slotted" and don't apply the backoff to them.

**Comment by alrevuelta** [comment]
@jm-clius Fixed your comment by adding a flag to make respecting the backoff optional. Let me know what you think :)

This feature is currently unused and will be used in https://github.com/waku-org/nwaku/issues/1477 in the `connectivity loop` that you also refer to, and only for relay peers.

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (jm-clius) [inline waku/v2/node/peer_manager/peer_manager.nim:83]: Afaict this will not allow any differentiation between "still backing off" and other connection failures. Would this be something applications require? In other words, this would not/should not stop the application from attempting to dial this peer continuously as there is no indication that under the hood we're still backing off?

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (jm-clius) [inline waku/v2/node/peer_manager/peer_manager.nim:96]: Any reason not to use `dialFut.withTimeout()`?

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (jm-clius) [inline waku/v2/node/peer_manager/peer_manager.nim:116]: This seems a bit weird to me - the fact that me dialing some peer could result in it being removed from the peer store (and doing it again would result in it being added again, presumably?). Of course, we know that for the _application_ of "continuously attempt to connect to all available `relay` peers" this would make sense, i.e. to eventually stop attempting to dial peers that continue to fail. But this highlights my concern that this logic should not be part of the dialer, even if behind a `respectBackoff` flag.

**Review by jm-clius — COMMENTED** [review]
Thanks! I will review in more detail after the weekend, in case I've missed some intricacies. :)

Adding the `respectBackoff` flag is indeed better, but I'm still not convinced that this logic should be part of the dialer.

As I see it, we have two distinct needs here:
1. A dialer interface that must allow any protocol/application to attempt to dial peers, maintain connectivity related peer books, etc.
2. At least one "application" of the peer manager/store (the "connect loop") that wants to use this dialer to continuously attempt to connect to all relevant peers in the peer store. This application should only attempt connection to peers that are not being backed off from, remove peers that it has the authority to do (i.e. not static peers), etc.

Mixing logic from (2) into (1) seems to me to create some confusion. What if an application wants to `respectBackoff`? Currently it will simply receive a `none(Connection)` in return if the peer is being backed off from. In my mind it has no reason not to continue attempting to connect to this peer and doesn't gain any information to help it make better decisions in future. Furthermore, the dialer will now make decisions such as removing peers from the peer store after max failed attempts. It seems to me to be doing that because it "knows" that this is what the connect loop would expect of it, since the connect loop is selecting peers from the peer store for attempted connection. This has no bearing, however, on other applications/protocols that may be managing their own peers.

**Comment by alrevuelta** [comment]
@jm-clius I agree with your 1. and 2. needs. Main reason I added the `respectBackoff` to `dialPeer` and mixed the logic is so that its easier to unit test. If I respect the backoff directly in the loop, that makes it more difficult to unit test.

Plan b is to have a separate function, but that involves duplicating lots of logic: update metrics, dial + timeouts, etc.

Will convert the PR to draft and implement a vanila connectivity loop in a separate PR. Then I will add the backoff to that connectivity loop, which should comply with your "not mixing logic" requirement.

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim]: just reverted my changes, using again withTimeout.

i thought there was a possible race condition and wanted also to cancel the dial if the timer timed out, but noticed it didnt made much sense.

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim]: fixed

File: waku/v2/node/waku_node.nim
Comment (rymnc) [inline waku/v2/node/waku_node.nim:911]:
```suggestion
    if discoveredPeersRes.isOk():
```
nit, but i believe this is the style guide we're adopting

Comment (rymnc) [reply waku/v2/node/waku_node.nim:911]: maybe we should handle an error if `findRandomPeers` fails. An error log atleast

File: waku/v2/node/peer_manager/waku_peer_store.nim
Comment (rymnc) [inline waku/v2/node/peer_manager/waku_peer_store.nim:94]:
```suggestion
  let lastFailed = peerStore.getLastFailedPeer(peerId)
```
or similar, which would allow us to change the underlying datastructure if required in the futute

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (rymnc) [inline waku/v2/node/peer_manager/peer_manager.nim:41]: Just wondering what people's views are on making this configurable by the operator. To me, it allows for more aggressive dialing behaviour. wdyt?

**Review by rymnc — COMMENTED** [review]
LGTM, just a couple questions :)

File: waku/v2/node/waku_node.nim
Comment (alrevuelta) [reply waku/v2/node/waku_node.nim:911]: sure thanks!

File: waku/v2/node/peer_manager/waku_peer_store.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/waku_peer_store.nim:94]: I agree with this, but I'm just trying to follow nimlibp2p peerstore https://github.com/status-im/nim-libp2p/blob/unstable/libp2p/peerstore.nim#L148  pattern, of using custom books.

Doing this will require to have 6-7 new getter functions, with just one line of code, and not sure I see the benefit right now.

But I'm totally open for suggestions.

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim:41]: mm these are some default safe values that i've tested. what do you mean by configurable? new cli flags? not sure if that would be to low level for an operator? note though that they can be changed when creating the peermanager.

File: waku/v2/node/peer_manager/waku_peer_store.nim
Comment (rymnc) [reply waku/v2/node/peer_manager/waku_peer_store.nim:94]: Sounds good! just something to keep in mind if we decide to have more complex functionality later.

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (rymnc) [reply waku/v2/node/peer_manager/peer_manager.nim:41]: yeah, I mean cli flags :)

File: waku/v2/node/peer_manager/waku_peer_store.nim
Comment (jm-clius) [inline waku/v2/node/peer_manager/waku_peer_store.nim:100]: Afaict, this is not yet used anywhere? Given the fact that in existing deployments peer IDs are cycled very often (I think), we should add a mechanism to manage the size of the peer store fairly urgently - unsure of the implication if this memory essentially leaks in the meantime.

**Review by jm-clius — APPROVED** [review]
Thanks! Makes sense to me now that the `dialPeer` mechanism does not make any decisions on whether to dial a peer or not. Some minor comments below. My biggest concern is prioritising some mechanism to manage the number of peers kept in the store to avoid leaks. Would be good to monitor peer management behaviour closely once this is merged (and auto-deployed to `wakuv2.test`)

File: waku/v2/node/peer_manager/peer_manager.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim:353]: ouch, leftover. will fix.

Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim:353]: ah right! will fix.

File: waku/v2/node/waku_node.nim
Comment (alrevuelta) [reply waku/v2/node/waku_node.nim:886]: The main change here is to use `switch.dial` instead of `peerManager.dialPeer` for the ping. Main reason is that `dialPeer` updates metrics on ok/nok connections, failed attempts, last time failed, Can/Cannot be connected etc.

And since here we are just pinging this peer, I don't think using that function makes sense. For example, we will be updating the metrics with ok connections, every time we ping a peer.

And actually this should be more like `getConnection` because we are not dialing/connecting to any peer, but getting an already existing connection and sending a ping.

I agree with explicir error handling, but here I had to add the `try` `except` because `switch.dial` doesn't return `Result[xx]`.

File: waku/v2/node/peer_manager/waku_peer_store.nim
Comment (alrevuelta) [reply waku/v2/node/peer_manager/waku_peer_store.nim:100]: Yep, not used in this PR but tracked here in "Prune peers from peerstore".
shouldn't leak since its limited by `.withPeerStore(capacity=xxx)` but yes, we have to handle it more gracefully.

**Review by LNSD — APPROVED** [review]
LGTM ✅

**Pull request merged**
